### PR TITLE
187129573 change teacher status

### PR DESCRIPTION
**Pull request by ArushC**
### [Pivotal Tracker Link][tracker]

[tracker]: https://www.pivotaltracker.com/story/show/your-story-id

## What this PR does:

Allows changing a teacher's application status. The frontend was demonstrated at the customer meeting.

This pull request fixes|implements (pick one...) ______.

#### Who authored this PR?

Arush

### How should this PR be tested?

Look at the Cucumber tests. They verify that the feature works as intended.

* Is there a deploy we can view?

Yes: [https://sp24-01-bjc-teachers-d0e972207d82.herokuapp.com/dashboard](https://sp24-01-bjc-teachers-d0e972207d82.herokuapp.com/dashboard)

* What do the specs/features test?
* Are there edge cases to watch out for?

### Checklist:

- [x] Has this been deployed to a staging environment or reviewed by a customer?
- [x] Tag someone for code review (either a coach / team member)
- [x] I have renamed the branch to match PivotTracker's suggested one (necessary for BlueJay) (e.g. `michael/12345-add-new-feature` Any branch name will do as long as the story ID is there. You can use `git checkout -b [new-branch-name]`)

### CODE REVIEW

**Comment by perryzjc** [comment]
Looks good to me as well. It seems to fix everything mentioned in Professor Ball's Initial comments.

If anything else is needed, I think we need some notes to document what admins need to do after the PR, especially about the migration.

Based on the discussions we had in today's lecture with Michael (Friday, Mar 8), Michael suggested we just document the manual steps to update the email template (since there are only quite a few templates right now), such as the values to fill out for the new `to` field.

We could possibly write something for the `Are there any complications to deploying this?` section as below:

1. Run "rails db:migrate" since there are schema changes  (a new `TO` field)
2. (Optional) Run `rails db:reseed` if you want to directly import the [updated email templates we made](https://github.com/beautyjoy/BJC-Teacher-Tracker/pull/292/files#diff-7e0436c5774a8f6f798727122bfc14661031feaed667529064143541b17d2a1d) (updated email content, new template, and the pre-filled `to` fields)
3. (Alternative way to step 2) Manually update the existing email template with the new content.
    - After running step 1 (migrate), for the new `to` field, fill in `{{teacher_email}}, {{teacher_personal_email}}` as the template value, or some default emails, e.g. `lmock@berkeley.edu, contact@bjc.berkeley.edu`
    - Also need to update the template content for `Denial Email` and `Request More Info Email`, as they no longer use the original `reason` field, but the `denial_reason` or `request_reason` instead. Sample format can be found at the `seed_data.rb` we provided, such as:
```
  @deny_email = <<-DENY_EMAIL1
    <p>Dear [Recipient's Name],</p>
    <p>Thank you for your email. We have received your message and we are sorry to inform you that your application has been rejected.</p>
    <p>Best regards,</p>
    <p>[Your Name]</p>
    <p>Below, you can find the reason as to why it was rejected </p>
    <p>
      {{ denial_reason | strip_tags}}
    </p>
  DENY_EMAIL1
```

File: app/controllers/email_templates_controller.rb [hunk 1]
```diff
@@ -14,8 +14,13 @@ def edit
   def update
     template = EmailTemplate.find(params[:id])
     template.update(template_params)
-    template.save!
-    redirect_to email_templates_path
+    if template.save
+      flash[:success] = "Updated #{template.title} template successfully."
+      redirect_to email_templates_path
+    else
+      flash[:alert] = "Failed to save #{template.title} template: " + template.errors.full_messages.join(", ")
```
Comment (fp456) [inline app/controllers/email_templates_controller.rb:21]: codecov: Added lines #L21 - L22 were not covered by tests

File: app/views/main/_deny_modal.html.erb [hunk 2]
```diff
@@ -5,23 +5,25 @@
...
-              </select>
-            </div> -->
-            <div class="form-check">
-              <%= check_box_tag :skip_email, 'skip_email', false, class: 'form-check-input' %>
-              <%= label_tag 'skip_email', 'Skip email notifcation?', class: 'form-check-label' %>
-            </div>
+          <%= hidden_field_tag :action_type, '', class: "form-control action-type" %>
+          <%= label_tag 'action_reason', 'Reason', class: 'control-label' %>
+          <%= text_field_tag :action_reason_placeholder, '', class: "form-control action-reason", id: "action_reason" %>
+          <!-- <div class="form-group">
```
Comment (fp456) [inline app/views/main/_deny_modal.html.erb:14]: clean up: remove commented lines

File: app/controllers/email_templates_controller.rb (hunk 1 above)
Comment (ArushC) [reply app/controllers/email_templates_controller.rb:21]: good catch, I forgot to test the happy path. Added a test case to features/email_template.feature in the latest commit

File: app/mailers/teacher_mailer.rb [hunk 3]
```diff
@@ -48,7 +52,9 @@ def liquid_assigns
     base_rules = {
       bjc_password: Rails.application.secrets[:bjc_password],
       piazza_password: Rails.application.secrets[:piazza_password],
-      reason: @reason
+      # TODO: Review if below two are needed, or can they be refractored?
```
[the remark this answers is not in the document]
Comment (ArushC) [reply app/mailers/teacher_mailer.rb]: resolved in latest commit

File: app/views/main/_deny_modal.html.erb (hunk 2 above)
Comment (ArushC) [reply app/views/main/_deny_modal.html.erb]: resolved in latest commit

File: app/controllers/main_controller.rb [hunk 4]
```diff
@@ -12,8 +12,10 @@ def index
       redirect_to pages_path, success: "Welcome back, #{current_user.first_name}!"
     # if this teacher is not validated (not_reviewed or info_needed), redirect to edit
     elsif is_teacher? && !current_user.validated?
-      redirect_to edit_teacher_path(current_user.id),
-                 alert: "Your applicating is currently #{current_user.application_status}. You may update your information."
+      message = "Your application is #{mapped_application_status(current_user.application_status)}. You may update your information. Please check your email for more information."
+      # Both `info_needed` and `not_reviewed` flash message should be flash[:warn] style
+      # flash[:warn] for yellow warning message in bootstrap style
```
Comment (fp456) [inline app/controllers/main_controller.rb:17]: You can delete the commented line

File: app/controllers/main_controller.rb [hunk 5]
```diff
@@ -12,8 +12,10 @@ def index
       redirect_to pages_path, success: "Welcome back, #{current_user.first_name}!"
     # if this teacher is not validated (not_reviewed or info_needed), redirect to edit
     elsif is_teacher? && !current_user.validated?
-      redirect_to edit_teacher_path(current_user.id),
-                 alert: "Your applicating is currently #{current_user.application_status}. You may update your information."
+      message = "Your application is #{mapped_application_status(current_user.application_status)}. You may update your information. Please check your email for more information."
```
Comment (fp456) [inline app/controllers/main_controller.rb:15]: needs testing

File: app/controllers/teachers_controller.rb [hunk 6]
```diff
@@ -127,10 +130,23 @@ def update
     redirect_to edit_teacher_path(current_user.id), notice: "Successfully updated your information"
   end
 
+  def send_email_if_application_status_changed_and_email_resend_enabled
+    if @teacher.application_status_changed? && params[:skip_email] == "No"
+      case @teacher.application_status
```
Comment (fp456) [inline app/controllers/teachers_controller.rb:135]: Missing testing for the case.

**Review by fp456 — CHANGES_REQUESTED** [review]
Nice job with the testing. I know it was a large PR but there were just some missing tests that it seems you added. Just minor things but overall, good job. I reviewed this PR in 2 separate days so I just made a final comment about adding tests for the case statement inside the send_email_if_application_status_changed_and_email_resend_enabled method. If it's more convenient, you can add the tests here if not, in a separate PR would be better since this PR is already large. In the future, let's try to keep the PRs shorter for a faster revision and also it'll help you break things down. Good job again!

File: app/controllers/main_controller.rb (hunk 4 above)
Comment (cycomachead) [inline app/controllers/main_controller.rb:17]:
```suggestion
```

File: app/controllers/main_controller.rb [hunk 7]
```diff
@@ -27,4 +29,14 @@ def dashboard
...
     @schools = School.validated.order(teachers_count: :desc).limit(20)
   end
+
+  private
+  def mapped_application_status(status)
+    status_map = {
+      "info_needed" => "requested to be updated",
+      "not_reviewed" => "not reviewed"
+    }
+    # Fetch the mapped status if present; otherwise, return the original status
```
Comment (cycomachead) [inline app/controllers/main_controller.rb:39]:
```suggestion
```

File: app/controllers/main_controller.rb [hunk 8]
```diff
@@ -27,4 +29,14 @@ def dashboard
     @statuses = Teacher.validated.group(:status).order(count_all: :desc).count
     @schools = School.validated.order(teachers_count: :desc).limit(20)
   end
+
+  private
+  def mapped_application_status(status)
+    status_map = {
+      "info_needed" => "requested to be updated",
+      "not_reviewed" => "not reviewed"
+    }
```
Comment (cycomachead) [inline app/controllers/main_controller.rb:38]:
```suggestion
    }.with_indifferent_access
```

This allows a hash to treat symbols and strings as equivalent keys.

File: app/views/teachers/_form.html.erb [hunk 9]
```diff
@@ -113,4 +186,7 @@
     }
 
   }
+
+
+
```
Comment (cycomachead) [inline app/views/teachers/_form.html.erb:191]:
```suggestion
```

File: db/migrate/20240218005604_add_to_field_to_email_template.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+class AddToFieldToEmailTemplate < ActiveRecord::Migration[6.1]
+  def change
+    add_column :email_templates, :to, :string
+    #below is the default prepopulated 'to' field value,
+    to_field = "{{teacher_email}}, {{teacher_personal_email}}"
+    EmailTemplate.update_all(to: to_field)
```
Comment (cycomachead) [inline db/migrate/20240218005604_add_to_field_to_email_template.rb:6]:
```suggestion
```

Remember, we don't want to include data changes in the migrations files.

**Review by cycomachead — APPROVED** [review]
🎉
